src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/MemorySegment.java

```diff
@@ -836,27 +888,730 @@ static MemorySegment mapFile(Path path, long bytesOffset, long bytesSize, FileCh
+     * @throws IllegalStateException if the scope associated with this segment has been closed, or if access occurs from
+     * a thread other than the thread owning that scope.
+     * @throws IndexOutOfBoundsException when the dereference operation falls outside the <em>spatial bounds</em> of the
+     * memory segment.
+     */
+    @ForceInline
+    default MemoryAddress getAtIndex(ValueLayout.OfAddress layout, long index) {
+        return (MemoryAddress)layout.accessHandle().get(this, Utils.scaleOffset(this, index, layout.byteSize()));
+    }
+
+    /**
+     * Writes an address to this segment and index with given layout.
+     *
+     * @param layout the layout of the memory region to be written.
+     * @param index index (relative to this segment). For instance, if this segment is a {@link #isNative()} segment,
+     *               the final address of this write operation can be expressed as {@code address().toRowLongValue() + (index * layout.byteSize())}.
+     * @param value the byte value to be written.
+     * @throws IllegalStateException if the scope associated with this segment has been closed, or if access occurs from
+     * a thread other than the thread owning that scope.
+     * @throws IndexOutOfBoundsException when the dereference operation falls outside the <em>spatial bounds</em> of the
+     */
+    @ForceInline
+    default void setAtIndex(ValueLayout.OfAddress layout, long index, Addressable value) {
+        layout.accessHandle().set(this, Utils.scaleOffset(this, index, layout.byteSize()), value.address());
+    }
+
+
+    /**
+     * Copies a number of elements from a source segment to a destination array,
+     * starting at a given segment offset (expressed in bytes), and a given array index, using the given source element layout.
+     * @param srcSegment the source segment.
+     * @param srcLayout the source element layout. If the byte order associated with the layout is
+     * different from the native order, a byte swap operation will be performed on each array element.
+     * @param srcOffset the starting offset, in bytes, of the source segment.
+     * @param dstArray the destination array.
+     * @param dstIndex the starting index of the destination array.
+     * @param elementCount the number of array elements to be copied.
+     * @throws  IllegalArgumentException if {@code dstArray} is not an array, or if it is an array but whose type is not supported.
+     * Supported array types are {@code byte[]}, {@code char[]},{@code short[]},{@code int[]},{@code float[]},{@code long[]} and {@code double[]}.
+     */
+    @ForceInline
+    static void copy(
+            MemorySegment srcSegment, ValueLayout srcLayout, long srcOffset,
+            Object dstArray, int dstIndex, int elementCount) {
+        Objects.requireNonNull(srcSegment);
+        Objects.requireNonNull(dstArray);
+        Objects.requireNonNull(srcLayout);
+        long baseAndScale = getBaseAndScale(dstArray.getClass());
```

Comment:
Should we not add a check that the srcLayout's carrier type fits the dstArray type?